Issue 17 create op for stream

vladislav-kulakovsky opened the pull request.

vladislav-kulakovsky committed:
add invokePathCreateStream function, add code to CREATE label

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

"Signed-off-by: Vladislav Kulakovsky kulakovsky.vladislav@gmail.com"

vladislav-kulakovsky committed:
Merge branch 'master' into issue-17-CreateOpForStream
Merge conflicts fixed

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

"Signed-off-by: Vladislav Kulakovsky kulakovsky.vladislav@gmail.com"

vladislav-kulakovsky committed:
add .gradle to gitignore;
scopeMap replaced to streamMap and new scopeMap;
invokePathCreateStream is completed

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

"Signed-off-by: Vladislav Kulakovsky kulakovsky.vladislav@gmail.com"

vladislav-kulakovsky committed:
add operation status when if invokePathCreateStream exit with fail;
temporarily changed the check in READ block, utnil the architecture problem is resolved

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

"Signed-off-by: Vladislav Kulakovsky kulakovsky.vladislav@gmail.com"

vladislav-kulakovsky committed:
scopeMap structure changed back to <string, map<string>>;
scopeStream removed;
invokePathCreateStream changed according to changes above

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

"Signed-off-by: Vladislav Kulakovsky kulakovsky.vladislav@gmail.com"

vladislav-kulakovsky committed:
merge conflict solved;
function invokePathCreateStream renamed to invokePathCreate

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

"Signed-off-by: Vladislav Kulakovsky kulakovsky.vladislav@gmail.com"

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -226,7 +231,57 @@ private void invokeDirectoryNio(PathOperation<? extends PathItem> pathOperation)
+		final String scopeName = path.substring(0, path.indexOf("/"));
+		final String streamName = path.substring(path.indexOf("/")+1);
+
+		final Function<String, Map<String, String>> createScopeComputeIfAbsent = (final String k) -> {
```
let's convert this function to a class' method

vladislav-kulakovsky replied:
@akurilov, If I make this function as a class method then I need to create the local object of StreamManager

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -226,7 +231,57 @@ private void invokeDirectoryNio(PathOperation<? extends PathItem> pathOperation)
+			return new ConcurrentHashMap<String, String>();
+		};
+
+		final Function<String, String> createStreamComputeIfAbsent = (final String k) -> {
```
the same as above: please convert this to a class' method

vladislav-kulakovsky replied:
the same problem with StreamManager as above comment and in addition 'scopeName' variable will be not defined cause it computes in invokePathCreate

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -226,7 +231,57 @@ private void invokeDirectoryNio(PathOperation<? extends PathItem> pathOperation)
+				return false;
+			}
+		} else {
+			Loggers.MSG.info(
```
let's ignore the return value and continue to not to duplicate these branches

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -226,7 +231,57 @@ private void invokeDirectoryNio(PathOperation<? extends PathItem> pathOperation)
+			return k;
+		};
+
+		if (null != scopeMap.computeIfAbsent(scopeName,	createScopeComputeIfAbsent)) {
```
this condition will be `true` always because the map (new either existing previously) will be returned

vladislav-kulakovsky replied:
Does it mean that then we generally need to throw out the condition and just call the function?

akurilov replied:
not just call but save the result into the local variable in order to not invoke `scopeMap.get` after again

vladislav-kulakovsky committed:
Conditions removed from invokeCreatePath function

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

"Signed-off-by: Vladislav Kulakovsky kulakovsky.vladislav@gmail.com"
